throttle shell output to maintain responsiveness

justinmk opened the pull request:
Closes #1234 
- Commit 1 (resetting the TUI event queue if it floods) is the main safeguard against #1234. But it's quite an edge case because `yes` produces tons (GB in just a few minutes) of data very quickly.
- Commit 2 (skipping `:!` spam periodically) is a more general "cheat" that works for all UIs and greatly improves responsiveness in the TUI even when it is "spammed" but not "flooded".

Try these shell-based ex-commands before/after this PR:

``` vim
:grep -r '' *
:make
:!yes
:!grep -r '' *
:!git grep ''
:!cat foo
:!echo foo
:!while true; do date; done
:!for i in `seq 1 20000`; do echo XXXXXXXXXX $i; done
```
- Output will be periodically skipped and a pulsing "..." displayed. In all cases the last few lines of the command should always be shown (these are usually what the user cares about, if anything).
- Data is not lost, so `:read` still reads all data; `:make` and `:grep` still send all lines to the quickfix list; etc.

Output from `:!` exceeding ~10 KB is not useful to users. This is even more true for nvim because:
- `:!` and `system()` use pipes, the nvim built-in pager doesn't work, so any output that scrolled off the screen is gone
- `:!` and `system()` use pipes, interactive programs like `:!git grep ''` won't treat the output as tty, and won't page themselves (unlike Vim)
- This PR throttles output greater than ~10 KB. 
  - Significant perceived performance improvement in terminal-based Nvim.
  - `:read !foo` and `:<range>!foo` still correctly receive all data; throttling only affects UI.
  - `execute('!foo')` behavior did not change. (Added a test for this.)
- Future work (?): For user-initiated `:!` invocations, use `:terminal` instead?
  - This will be hard to make "perfect", but would be a valuable feature if possible.
# Note

The same "TUI flooding" problem can be triggered by doing 

```
:g/^/p
```

on a very large file and then pressing `G` to scroll to the bottom of the messages in the pager. This PR doesn't address `:g` (or anything that uses the built-in pager); this PR only address shell-command output.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,19 +309,44 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+      if (*p == NL) {
+        lines++;
+      }
+    }
```
Could be replaced with `memcnt()`.

justinmk replied:
@oni-link Thanks! Instead of counting lines, I'm thinking some upper limit of "bytes seen" for a given `source` should trigger the throttling.

justinmk replied:
An idea for a more sophisticated "throttle trigger", might be the keep a count of the pending `thread_events` per-queue and then maybe ignore them until processing catches up. 

But maybe the arbitrary limit  (this PR) is good enough: it's probably not useful for anyone to see more than 1000 lines blasted in the pager, in _any_ UI (TUI or GUI). So the approach I was thinking here was to "snip" the middle of long outputs, and print the last N lines (as well as the first N lines).

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,19 +309,44 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+      remaining = 1;
+      if (last_col >= Columns) {
+        last_col = 0;
+        skip_msg[1] = '\n';
```
Where would this field be read?

oni-link commented on src/nvim/tui/tui.c:
```diff
@@ -583,6 +588,17 @@ static void tui_flush(UI *ui)
+  int nrevents = queue_count(data->loop->thread_events);
+  // QUEUE *q = QUEUE_HEAD(&queue->headtail);
+  // while(true) {
+  // }
```
Will this be removed?

justinmk replied:
Yes, this PR is very much unfinished, though I will send updates tonight.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -583,6 +588,17 @@ static void tui_flush(UI *ui)
+  // QUEUE *q = QUEUE_HEAD(&queue->headtail);
+  // while(true) {
+  // }
+  ILOG("thread_events=%d", nrevents);
```
Yep, I'll add a condition or remove this. `yes` can generate MB of data in a couple seconds, and GB in just a few minutes.

justinmk committed:
os/shell: Throttle :! output, pulse "..." message.

Periodically skip :! spam. This is a "cheat" that works for all UIs and greatly
improves responsiveness when :! spams MB or GB of output:
    :!yes
    :!while true; do date; done
    :!git grep ''
    :grep -r '' *

After ~10KB of data is seen from a single :! invocation, output will be skipped
for ~1s and three dots "..." will pulse in the bottom-left. Thereafter the
behavior alternates at every:
    * 10KB received
    * ~1s throttled

This also avoids out-of-memory which could happen with large :! outputs.

Note: This commit does not change the behavior of execute(':!foo').
      execute(':!foo') returns the string ':!foo^M', it captures *only* Vim
      messages, *not* shell command output. Vim behaves the same way.
      Use system('foo') for capturing shell command output.

Closes #1234

Helped-by: oni-link <knil.ino@gmail.com>

justinmk committed:
tui: "backpressure": Drop messages to avoid flooding.

Closes #1234

multiqueue:
- Implement multiqueue_size()
- Rename MultiQueueItem.parent to MultiQueueItem.parent_item, to avoid confusion
  with MultiQueue.parent.

justinmk commented:
Rebased, RFC. Reviews appreciated.

Not planning to write new tests for this because it would be very slow. Maybe I should try though.

justinmk commented:
Rebased and add a test.

justinmk commented on test/functional/ui/output_spec.lua:
```diff
@@ -39,4 +40,17 @@ describe("shell command :!", function()
+      XXXXXXXXXX                                        |
+      ..                                                |
+      {10:Press ENTER or type command to continue}{1: }          |
+      {3:-- TERMINAL --}                                    |
```
This test is a bit fragile, I'm not sure only two dots will appear on all platforms, it depends on the buffer size.

If this is too fragile, I may need to add a function to `screen.lua` that checks for a pattern without asserting the exact screen state.

justinmk commented on test/functional/ui/output_spec.lua:
```diff
@@ -39,4 +40,11 @@ describe("shell command :!", function()
+    child_session.feed_data(
+      ":!for i in $(seq 2 3000); do echo XXXXXXXXXX; done\n")
+    -- If a line with only a dot "." appears, then throttling was triggered.
+    screen:expect("\n.", nil, nil, nil, true)
```
Passing `any=true` to `screen:expect` checks that the string appears "somewhere" in the screen state. The expected string `"\n."` means that throttling was triggered for the shell output.

justinmk commented on test/functional/ui/screen.lua:
```diff
@@ -207,7 +207,15 @@ function Screen:try_resize(columns, rows)
+-- condition:   Function asserting some arbitrary condition.
+-- any:         true: Succeed if `expected` matches ANY screen line(s).
+--              false (default): `expected` must match screen exactly.
+function Screen:expect(expected, attr_ids, attr_ignore, condition, any)
```
Added `any` parameter.

justinmk commented on test/functional/ui/screen.lua:
```diff
@@ -229,21 +237,34 @@ function Screen:expect(expected, attr_ids, attr_ignore, condition)
+          return (
+            'Row ' .. tostring(i) .. ' did not match.\n'
+            ..'Expected:\n  |'..table.concat(msg_expected_rows, '|\n  |')..'|\n'
+            ..'Actual:\n  |'..table.concat(actual_rows, '|\n  |')..'|\n\n'..[[
```
Modified the Expected/Actual messages to indent the contents for readability.

jamessan reviewed (approved):
LGTM

justinmk committed:
os/shell: do_os_system(): Always show last chunk.

This ameliorates use-cases like:
    :!cat foo.txt
    :make
where the user is interested in the last few lines of output.

Try these shell-based ex-commands before/after this commit:
    :grep -r '' *
    :make
    :!yes
    :!grep -r '' *
    :!git grep ''
    :!cat foo
    :!echo foo
    :!while true; do date; done
    :!for i in `seq 1 20000`; do echo XXXXXXXXXX $i; done

In all cases the last few lines of the command should always be shown,
regardless of where throttling was triggered.

justinmk commented:
Added a commit which ensures that the last chunk is always displayed. Still need to update the test and docs.

oni-link commented:
If I use `:!yes` followed by `<C-C> :!cat README.md`, it shows me output from `yes`.

justinmk commented:
Thanks. Yes the init needs to be fixed.

justinmk commented:
Updated. This works well for [all cases I can think of](https://github.com/neovim/neovim/pull/5396#issue-179663300).

> If I use `:!yes` followed by `<C-C> :!cat README.md`, it shows me output from `yes`.

@oni-link That should be fixed now.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+///             output should be skipped. size=0 resets the internal state and
+///             returns the previous decision.
+///
+/// @returns true if output was skipped and pulse was displayed.
```
Is this not more like a suggestion for skipping output?

justinmk replied:
Updated the docstring, and renamed the functions in 91c16359b6574a897f69073f38f098b87dbf5882

justinmk replied:
Updated the docstring, and renamed the functions in 91c16359b6574a897f69073f38f098b87dbf5882

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+  } else if (size > 0) {
+    // Length of the old data that can be kept.
+    size_t old_keep_len = MAX_CHUNK_SIZE - size;
+    size_t old_keep_start = last_skipped_len - old_keep_len;
```
Is `old_keep_len` always smaller than `last_skipped_len`?

justinmk replied:
@oni-link Thanks for noticing that, see 07dc5592b2e0b952f9a8aa7d72001eefd073aeb9. This fixed behavior I noticed with certain output patterns.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+  // Pulse "..." at the bottom of the screen.
+  size_t tick = (visit == max_visits)
+                ? 3  // Force all dots "..." on last visit.
+                : (visit + 1) % 4;
```
Is the `+1` still needed? The first pulse now starts with two dots.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
+    } else {
+      received = visit = 0;
```
Should `max_visits` also be reset?

justinmk replied:
I don't reset it until the next shell command. This is to avoid calls to hrtime() which is quite slow.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
 ///
-/// @param output       Data to append to screen lines.
+/// @param output       Data to append to screen lines, or NULL to append the
+///                     last throttled (skipped) output.
```
`output=NULL`  does not append throttled data.

justinmk replied:
Will push some local changes soon.

justinmk committed:
out_data_decide_throttle(): Avoid too-small final chunk.

justinmk commented:
If no more bugs are found I'll let this sit for a week, we'll release 0.1.6 before including this.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,141 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+                        : 0;
+    // Shift the kept part of the old data to the start.
+    memmove(last_skipped, last_skipped + keep_start, keep_len);
+    // Copy the entire new data to the remaining space.
```
This could be written a bit shorter:
```c
    size_t keep_len   = MIN(last_skipped_len, MAX_CHUNK_SIZE-size);
    size_t keep_start = last_skipped_len - keep_len;
    
    // Shift the kept part of the old data to the start.
    if (keep_start) {
      memmove(last_skipped, last_skipped + keep_start, keep_len);
    }
```
and making `OUT_DATA_THRESHOLD` unsigned to prevent a compiler warning because of the code change:
```c
#define OUT_DATA_THRESHOLD  1024 * 10U       // 10KB, "a few screenfuls" of data.
```

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,141 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+  }
+
+  if (output == NULL && size == SIZE_MAX) {   // Print mode
+    out_data_append_to_screen(last_skipped, last_skipped_len, false);
```
Should  `true` (EOF) be used here to start next command in column `0`?

justinmk replied:
I didn't notice a difference, but I will do so.

justinmk commented:
Rebased with suggested changes.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,140 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+      received = visit = 0;
+      return false;
+    }
+  }
```
If I use this script,
```bash
#!/bin/bash

# Produce enough output to reach the treshold
for i in {1..10000}
do
    echo "1: $i"
done

# Produce every second small chunk of output
for i in {1..1000}
do
    echo "2: $i"
    sleep 1
done
```
the only output I see is:
```
(cut out)
1: 1253
1: 1254
...   <- this is pulsing
```
With `max_visits` big it can take some time until  `visit>=max_visits` and output is shown.
Would have expected to see small chunks of output if they are slow.

justinmk replied:
Thanks for that test case. I changed it to timeout after 3 seconds instead of max_visits > 999.

To make it perfect, I guess 2 ring buffers would be needed. Hopefully we're close to "good enough". After looking more closely at the `:!` and `system()` implementations I'm convinced that we could change the implementation of `:!` to *always* use a pty/terminal job. So this throttling logic may not be needed after that (except on Windows, unless we get winpty working).

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -309,15 +318,140 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+    if (since < NS_1_SECOND) {
+      // Adjust max_visits based on the current relative performance.
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
```
Parentheses could be removed.

justinmk committed:
out_data_decide_throttle(): timeout instead of hard limit.

Instead of managing max_visits, check the time every N visits. This avoids edge
cases where max_visits is large but the chunk frequency slowed down, which would
causing the "..." pulse to show for a very long time. It's more important to
show output at reasonable intervals than to avoid calling os_hrtime().

justinmk commented:
Rebased to fix merge conflict. Only the last commit (a69bf55) contains changes.

justinmk commented:
https://github.com/neovim/neovim/pull/5396/commits/1a8baa34952d5f8d63dbccb433c66ea76348403f address @oni-link's [test case](https://github.com/neovim/neovim/pull/5396#discussion_r83535106) and is generally more flexible and lower-risk. Also re-tested this against [these cases](https://github.com/neovim/neovim/pull/5396#issue-179663300).

Going to merge this after CI. The `helpers.lua` enhancements are needed to fix the flaky `global_spec.lua` test.
